### PR TITLE
Transformation Fix for BoxClipper3D

### PR DESCRIPTION
**Pull request by gnedster**
Corrects the calculation for the point transformation for BoxClipper3D::clipPoint3D.

### CODE REVIEW

File: filters/include/pcl/filters/impl/box_clipper3D.hpp [hunk 1]
```diff
@@ -108,21 +108,23 @@ pcl::BoxClipper3D<PointT>::transformPoint (const PointT& pointIn, PointT& pointO
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 // ToDo use product on point.getVector3fMap () and transformatio_.col (i) to use the SSE advantages of eigen
+// http://www.opengl-tutorial.org/beginners-tutorials/tutorial-3-matrices/#an-introduction-to-matrices
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/impl/box_clipper3D.hpp]: Please remove this comment, I don't feel it's necessary here. Why do you felt it was particularly important to include it?

File: filters/include/pcl/filters/impl/box_clipper3D.hpp [hunk 2]
```diff
@@ -108,21 +108,23 @@ pcl::BoxClipper3D<PointT>::transformPoint (const PointT& pointIn, PointT& pointO
...
-                transformation_.data () [ 9]) <= 1 &&
-           fabs(transformation_.data () [ 1] * point.x +
-                transformation_.data () [ 4] * point.y +
-                transformation_.data () [ 7] * point.z +
-                transformation_.data () [10]) <= 1 &&
-           fabs(transformation_.data () [ 2] * point.x +
-                transformation_.data () [ 5] * point.y +
-                transformation_.data () [ 8] * point.z +
-                transformation_.data () [11]) <= 1 );
+  Eigen::Matrix4f mat = transformation_.matrix();
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/impl/box_clipper3D.hpp]: Don't make a copy of the matrix, just take a reference to it
```cpp
const Eigen::Matrix4f& mat = transformation_.matrix(); 
```

File: filters/include/pcl/filters/impl/box_clipper3D.hpp [hunk 3]
```diff
@@ -108,21 +108,23 @@ pcl::BoxClipper3D<PointT>::transformPoint (const PointT& pointIn, PointT& pointO
...
+                mat(0,2) * point.z +
+                mat(0,3)) <= 1 &&
+           fabs(mat(1,0) * point.x +
+                mat(1,1) * point.y +
+                mat(1,2) * point.z +
+                mat(1,3)) <= 1 &&
+           fabs(mat(2,0) * point.x +
+                mat(2,1) * point.y +
+                mat(2,2) * point.z +
+                mat(2,3)) <= 1 );
```
Comment (SergioRAgostinho) [inline filters/include/pcl/filters/impl/box_clipper3D.hpp]: Replace this entire block for 
```cpp
Vector4fMapConst pt = point.getVector4fMap ();
return (std::abs (mat.row (0) * pt) <= 1 && std::abs (mat.row (1) * pt) <= 1 && std::abs (mat.row (2) * pt) <= 1);
```
You can make multiline if for easier readability.

Note: I'm assuming point.data[3] to be equal to 1, as it is normally initialized.

**Comment by gnedster** [comment]
Hey @SergioRAgostinho, thanks for taking a look at the pull request. Some days later I stumbled upon the CropBox filter, which covers my use case for BoxClipper3D but with a much more intuitive API. 

I stumbled upon this issue when attempting to clip points in the z-axis in which it did not respect the arguments supplied at all, amongst other issues. Moreover the issue with this filter is that the transformation is applied to the point clouds rather than the unit cube, which leads to inputs like 1.0f/1000 to clip point boundaries of 1000. I suggest that the API is altered so that the arguments are more intuitive much in the way of CropBox at some point, but I'll limit the scope of my changes in this pull request to apply the correct transformation. 

Also is there a unit test example/standard that I can follow?

**Comment by gnedster** [comment]
Update - I've located the test_filter.cpp file and will add a test for BoxClipper3D there.

**Comment by SergioRAgostinho** [comment]
> Moreover the issue with this filter is that the transformation is applied to the point clouds rather than the unit cube, which leads to inputs like 1.0f/1000 to clip point boundaries of 1000.

It's conforming to the documentation so that's ok. http://docs.pointclouds.org/trunk/classpcl_1_1_box_clipper3_d.html#details

> I suggest that the API is altered so that the arguments are more intuitive much in the way of CropBox at some point, but I'll limit the scope of my changes in this pull request to apply the correct transformation.

But for that we have the CropBox, right?

> Update - I've located the test_filter.cpp file and will add a test for BoxClipper3D there.

I would actually suggest creating a new file for it. That test_filter.cpp is ultra packed with tests. Call it test_clipper.cpp and we'll aggregate there tests there for Clipper derived classes.

@taketwo I'm wondering why CropBox is also not deriving from Clipper, working as a particular case of BoxClipper3D 😅

**Comment by taketwo** [comment]
>  I'm wondering why CropBox is also not deriving from Clipper, working as a particular case of BoxClipper3D

One of PCL mysteries... Along with the question why `PPFEstimation` uses PFH features :laughing:

**Comment by taketwo** [comment]
Hey @gnedster, do you think you will have some spare to finish this soon? We have other people stumbling upon this issue and it's a pity that an almost complete solution is just lying around without being merged.

**Comment by gnedster** [comment]
Hey @taketwo, I've got the fix but I'm having trouble putting running the tests. Any guidance here?

**Comment by taketwo** [comment]
As Sergio suggested, create a new file in "test/filters", move the crop box tests there, and also add a test case for the `BoxClipper3D`. It does not have to cover all of its functionality. It would already be great if you just add a test that fails with current code and succeeds with your patch.

**Comment by gnedster** [comment]
I'm not so much concerned about file structure - more about the ability to run the test itself (i.e. a command line to make/run the test).

**Comment by taketwo** [comment]
Tests are disabled by default, so you need to pass `-DBUILD_global_tests=ON` when configuring with CMake (or toggle corresponding switch in CMake GUI). This will create a new target called `tests`, so executing `make tests` will build all the tests and then run them. But since you are interested in a single test, it may be quicker to just build it with `make test_filters` and run with `./test/filters/test_filters`.

**Comment by gnedster** [comment]
Should be good to go when tests pass. I added a variation of the test coverage made for CropBox while covering the degenerate case I stumbled upon in the first place (IIRC).

Cheers

**Comment by taketwo** [comment]
Thanks! One last small request: can you please format according to PCL style guide. Specifically, add spaces between the function names and brackets in function calls. One done, we are good to merge.

**Comment by gnedster** [comment]
Ok done.

**Pull request merged**